keyboards/crkbd/keymaps/antosha417/keymap.c

```diff
@@ -0,0 +1,429 @@
+const uint16_t PROGMEM bspcq_combo[] = {KC_I, J_CTLQ, COMBO_END};
+const uint16_t PROGMEM bspcwq_combo[] = {L_GUIQ, F_CTLQ, COMBO_END};
+const uint16_t PROGMEM saveq_combo[] = {S_GUIQ, J_CTLQ, COMBO_END};
+
+combo_t key_combos[] = {
+    [RU_COMBO] = COMBO(ru_combo, RUS_LANG),
+    [EN_COMBO] = COMBO(en_combo, EN_LANG),
+    [HEB_COMBO] = COMBO(heb_combo, HEB_LANG),
+    [ESC_COMBO] = COMBO(esc_combo, KC_ESC),
+    [TAB_COMBO] = COMBO(tab_combo, KC_TAB),
+    [ENT_COMBO] = COMBO(ent_combo, KC_ENT),
+    [DEL_COMBO] = COMBO(del_combo, KC_DEL),
+    [BSPC_COMBO] = COMBO(bspc_combo, KC_BSPC),
+    [SAVE_COMBO] = COMBO(save_combo, VIM_SAVE),
+    [BSPCW_COMBO] = COMBO(bspcw_combo, DELETE_WORD),
+
+    [RUQ_COMBO] = COMBO(ruq_combo, RUS_LANG),
+    [ENQ_COMBO] = COMBO(enq_combo, EN_LANG),
+    [HEBQ_COMBO] = COMBO(hebq_combo, HEB_LANG),
+    [ESCQ_COMBO] = COMBO(escq_combo, KC_ESC),
+    [TABQ_COMBO] = COMBO(tabq_combo, KC_TAB),
+    [ENTQ_COMBO] = COMBO(entq_combo, KC_ENT),
+    [DELQ_COMBO] = COMBO(delq_combo, KC_DEL),
+    [BSPCQ_COMBO] = COMBO(bspcq_combo, KC_BSPC),
+    [SAVEQ_COMBO] = COMBO(saveq_combo, VIM_SAVE),
+    [BSPCWQ_COMBO] = COMBO(bspcwq_combo, DELETE_WORD),
+};
+ 
+
+#ifdef OLED_ENABLE
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+  if (!is_keyboard_master()) {
+    return OLED_ROTATION_180;  // flips the display 180 degrees if offhand
+  }
+  return rotation;
+}
+
+#define L_QWERTY 2
+#define L_DVORAK 4
+#define L_LOWER 8
+#define L_RAISE 16
+#define L_TOP 32
+#define L_BOTTOM 64
+
+void oled_render_layer_state(void) {
+    oled_write_P(PSTR("layer: "), false);
+    if (layer_state < L_QWERTY) {
+      oled_write_ln_P(PSTR("qwerty"), false);
+    } else if (layer_state < L_DVORAK) {
+      oled_write_ln_P(PSTR("dvorak"), false);
+    } else if (layer_state < L_LOWER) {
+      oled_write_ln_P(PSTR("lower"), false);
+    } else if (layer_state < L_RAISE) {
+      oled_write_ln_P(PSTR("raise"), false);
+    } else if (layer_state < L_TOP) {
+      oled_write_ln_P(PSTR("top"), false);
+    } else if (layer_state < L_BOTTOM) {
+      oled_write_ln_P(PSTR("bottom"), false);
+    } else {
+      oled_write_ln_P(PSTR("unknown"), false);
+    }
+}
+
+
+char keylog_str[24] = {};
+
+const char code_to_name[60] = {
+    ' ', ' ', ' ', ' ', 'a', 'b', 'c', 'd', 'e', 'f',
+    'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p',
+    'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+    '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+    'R', 'E', 'B', 'T', '_', '-', '=', '[', ']', '\\',
+    '#', ';', '\'', '`', ',', '.', '/', ' ', ' ', ' '};
+
+void set_keylog(uint16_t keycode, keyrecord_t *record) {
+  char name = ' ';
+    if ((keycode >= QK_MOD_TAP && keycode <= QK_MOD_TAP_MAX) ||
+        (keycode >= QK_LAYER_TAP && keycode <= QK_LAYER_TAP_MAX)) { keycode = keycode & 0xFF; }
+  if (keycode < 60) {
+    name = code_to_name[keycode];
+  }
+
+  // update keylog
+  snprintf(keylog_str, sizeof(keylog_str), "%dx%d, k%2d : %c",
```

Comment:
just a heads up, snprintf uses around 1500 bytes of firmware space. It is recommended to not use it. 

`get_u8_str(number, ' ')` works a lot better, but is limiting to 3 characters (no more, no less).

Reply:
Thank you a lot for this comment! 
I just deleted this part since I do not really use it.